bridgepoint/prebuild.py

```diff
@@ -1005,11 +1013,21 @@ def accept_BinaryOperationNode(self, node):
         v_val_l = self.accept(node.left)
         v_val_r = self.accept(node.right)
 
+        l_s_dt = one(v_val_l).S_DT[820]()
+        l_s_irdt = one(l_s_dt).S_IRDT[17]()
         if node.operator.lower() in ['<', '<=', '==', '!=', '>=', '>',
                                      'and', 'or']:
             s_dt = self.s_dt('boolean')
+        elif node.operator.lower() in ['|', '+', '&', '^', '-'] and (l_s_irdt or
+            l_s_dt == self.s_dt('inst_ref<Object>') or l_s_dt == self.s_dt('inst_ref_set<Object>')):
+            # set the correct IRDT for set operation
+            if l_s_irdt is not None:
+                s_irdt = one(l_s_irdt).O_OBJ[123].S_IRDT[123](lambda sel: sel.isSet)
+                s_dt = one(s_irdt).S_DT[17]()
+            else:
+                s_dt = self.s_dt('inst_ref_set<Object>')
```

Comment:
So I simplified this block quite a bit. If the l_val type is an instance of S_IRDT, simply set the type of the binary operation to be the set of the same type. If the l_val type is one of the generic instance types, set the type to be `inst_ref_set<Object>`. There is no need to select through the variable to the class instance as I was doing before since all selections will be given an S_IRDT type if it is available.